game-core/src/test/java/games/strategy/triplea/delegate/power/calculator/TotalPowerAndTotalRollsTest.java

```diff
@@ -1,33 +1,690 @@
+      final int[] dice = new int[totalAttacks];
+      for (int i = 0; i < totalAttacks; i++) {
+        dice[i] = dieHit;
+      }
+
+      return TotalPowerAndTotalRolls.getTotalAaPowerThenHitsAndFillSortedDiceThenIfAllUseSameAttack(
+          dice, sortedDie, false, unitPowerAndRollsMap, validTargets, gameData, true);
+    }
+
+    @Test
+    void singleAaWithOneRollNoHit() {
+      final GameData gameData = givenGameData().withDiceSides(6).build();
+      final Unit unit = givenUnit("test", gameData);
+      unit.getUnitAttachment().setOffensiveAttackAa(2).setMaxAaAttacks(1);
+      final List<Unit> units = List.of(unit);
+      final List<Die> sortedDie = new ArrayList<>();
+
+      final Triple<Integer, Integer, Boolean> result =
+          whenGetPowerHitsResult(gameData, units, sortedDie, 6, 4);
+
+      assertThat(result, is(Triple.of(2, 0, true)));
+      assertThat(sortedDie, is(List.of(new Die(6, 2, Die.DieType.MISS))));
+    }
+
+    @Test
+    void singleAaWithTwoRoll() {
+      final GameData gameData = givenGameData().withDiceSides(6).build();
+      final Unit unit = givenUnit("test", gameData);
+      unit.getUnitAttachment().setOffensiveAttackAa(2).setMaxAaAttacks(2);
+      final List<Unit> units = List.of(unit);
+      final List<Die> sortedDie = new ArrayList<>();
+
+      final Triple<Integer, Integer, Boolean> result =
+          whenGetPowerHitsResult(gameData, units, sortedDie, 1, 4);
+
+      assertThat(result, is(Triple.of(4, 2, true)));
+      assertThat(
+          sortedDie, is(List.of(new Die(1, 2, Die.DieType.HIT), new Die(1, 2, Die.DieType.HIT))));
+    }
+
+    @Test
+    void twoAaWithSamePower() {
+      final GameData gameData = givenGameData().withDiceSides(6).build();
+      final Unit unit = givenUnit("test", gameData);
+      unit.getUnitAttachment().setOffensiveAttackAa(2).setMaxAaAttacks(1);
+      final Unit unit2 = givenUnit("test2", gameData);
+      unit2.getUnitAttachment().setOffensiveAttackAa(2).setMaxAaAttacks(1);
+      final List<Unit> units = List.of(unit, unit2);
+      final List<Die> sortedDie = new ArrayList<>();
+
+      final Triple<Integer, Integer, Boolean> result =
+          whenGetPowerHitsResult(gameData, units, sortedDie, 1, 4);
+
+      assertThat(result, is(Triple.of(4, 2, true)));
+      assertThat(
+          sortedDie, is(List.of(new Die(1, 2, Die.DieType.HIT), new Die(1, 2, Die.DieType.HIT))));
+    }
+
+    @Test
+    void twoAaWithDifferentPower() {
+      final GameData gameData = givenGameData().withDiceSides(6).build();
+      final Unit unit = givenUnit("test", gameData);
+      unit.getUnitAttachment().setOffensiveAttackAa(2).setMaxAaAttacks(1);
+      final Unit unit2 = givenUnit("test2", gameData);
+      unit2.getUnitAttachment().setOffensiveAttackAa(3).setMaxAaAttacks(1);
+
+      final List<Unit> units = List.of(unit, unit2);
+      final List<Die> sortedDie = new ArrayList<>();
```

Comment:
Out parameter here : (

This is something for us to remove and refactor. The support code is in part because it uses out parameters. Out parameters are certainly a C'ism and/or code written during java 1.4 when creating value objects was expensive. It is certainly more favored to now write a value object. As-is, the method under test is returning 4 different values, a Triple, and mutating this list parameter. One simplification we should strive to do is to remove the out parameters completely and have them be returned as values.